File: mailtidy/agent/legacy.py
"""遗留的"流水线 SOP 编排"实现。

这是 Phase 1 之前的 ``MailTidyAgent`` 类：把"获取邮件 -> LLM 判断 -> 套用策略
-> 生成执行计划 -> 执行 -> 出报告"几件事串成 SOP 工作流。该文件刻意保持薄：
所有真正的"决策智慧"都委托给：

- :mod:`mailtidy.llm`：稳定 LLM 抽象、路由、统计。
- :mod:`mailtidy.integrations.llm`：具体模型 / API / 本地运行时适配。
- :mod:`mailtidy.agent.policies`：哪些邮件做哪些动作、什么时候要确认。
- :mod:`mailtidy.integrations.email`：与真实 / Mock 邮箱的 IO 边界。
- :mod:`mailtidy.data.memory`：长期偏好与历史。

四个 SOP 对应四个 public 方法：

- ``run_cleanup``：收件箱清理 + 报告。
- ``draft_replies``：为 actionable 邮件批量草拟回复。
- ``daily_briefing``：早晨 2 分钟阅读的日报。
- ``scan_subscriptions``：订阅费扫描，导出 Markdown / CSV。

Phase 1 完成后，所有 SOP 会改写成 ``mailtidy.agent.loop`` 的 entry-point；
本类暂时保留以保证 CLI / 单测兼容。
"""

from __future__ import annotations

import re
from datetime import datetime

from mailtidy.agent.policies import DecisionPolicy
from mailtidy.data.memory import AgentMemory
from mailtidy.data.models import (
    ActionType,
    AgentPlan,
    Category,
    EmailMessage,
    ExecutionResult,
)
from mailtidy.data.reports import (
    cleanup_report,
    daily_brief,
    subscriptions_csv,
    subscriptions_markdown,
)
from mailtidy.integrations.email.base import EmailConnector
from mailtidy.llm.base import LLMClient


class MailTidyAgent:
    """负责把所有依赖装配起来并驱动 SOP。

    使用依赖注入的方式接收 ``connector`` / ``llm`` / ``memory`` / ``policy``，
    单测里可以用 Mock 完全跑通；接入真实 Gmail 时只要换两个实现即可。
    """

    def __init__(
        self,
        connector: EmailConnector,
        llm: LLMClient,
        memory: AgentMemory | None = None,
        policy: DecisionPolicy | None = None,
    ) -> None:
        self.connector = connector
        self.llm = llm
        # memory / policy 缺省时使用空记忆与默认阈值，方便临时跑 demo
        self.memory = memory or AgentMemory()
        self.policy = policy or DecisionPolicy()

    # ------------------------------------------------------------------ #
    # SOP 1：收件箱清理
    # ------------------------------------------------------------------ #
    def plan_cleanup(
        self,
        hours: int = 24,
        limit: int = 200,
        custom_dimensions: list[str] | None = None,
    ) -> tuple[AgentPlan, list[EmailMessage]]:
        """生成清理计划但不执行，便于 UI 先展示给用户审核。"""
        messages = self.connector.fetch_recent(hours=hours, limit=limit, unread_only=True)
        judgments = []
        for message in messages:
            judgment = self.llm.classify_email(message, custom_dimensions=custom_dimensions)
            judgments.append(self.policy.apply_memory(judgment, message.sender, self.memory))
        return self.policy.build_plan("inbox_cleanup", judgments), messages

    def execute_plan(self, plan: AgentPlan, auto_confirm: bool = False) -> ExecutionResult:
        """实际调用 connector 落地计划中的动作。"""
        result = ExecutionResult(processed=len(plan.judgments))
        for action in plan.actions:
            # 安全闸门：未拿到 auto_confirm，跳过所有需要确认的动作
            if action.requires_confirmation and not auto_confirm:
                result.skipped_confirmation += len(action.email_ids)
                continue
            if action.action == ActionType.ARCHIVE:
                self.connector.archive(action.email_ids)
                result.archived += len(action.email_ids)
            elif action.action == ActionType.LABEL and action.label:
                self.connector.label(action.email_ids, action.label)
                result.labeled += len(action.email_ids)
            elif action.action == ActionType.STAR:
                self.connector.star(action.email_ids)
                result.starred += len(action.email_ids)
            elif action.action == ActionType.MARK_READ:
                self.connector.mark_read(action.email_ids)
                result.marked_read += len(action.email_ids)
        return result

    def run_cleanup(
        self,
        hours: int = 24,
        limit: int = 200,
        custom_dimensions: list[str] | None = None,
        auto_confirm: bool = False,
    ) -> str:
        """完整的"清理 + 报告"端到端流程，CLI 直接调用这一个方法即可。"""
        plan, messages = self.plan_cleanup(hours=hours, limit=limit, custom_dimensions=custom_dimensions)
        result = self.execute_plan(plan, auto_confirm=auto_confirm)
        newsletters = [message for message in messages if self._category_for(plan, message.id) == Category.NEWSLETTER]
        newsletter_summary = self.llm.summarize_newsletters(newsletters)
        return cleanup_report(plan, result, messages, newsletter_summary)

    # ------------------------------------------------------------------ #
    # SOP 2：智能回复草拟
    # ------------------------------------------------------------------ #
    def draft_replies(self, email_ids: list[str] | None = None) -> ExecutionResult:
        """为 actionable 的邮件生成回复草稿，永远只写到草稿箱不发送。"""
        plan, messages = self.plan_cleanup(hours=24, limit=200)
        actionable = [
            message
            for message in messages
            if self._category_for(plan, message.id) == Category.ACTIONABLE
            and (email_ids is None or message.id in email_ids)
        ]
        result = ExecutionResult(processed=len(actionable))
        for message in actionable:
            draft = self.llm.draft_reply(message, self.memory.style_profile)
            self.connector.save_draft(message.id, draft)
            result.drafts_created += 1
        return result

    # ------------------------------------------------------------------ #
    # SOP 3：日报
    # ------------------------------------------------------------------ #
    def daily_briefing(self, custom_dimensions: list[str] | None = None) -> str:
        """早晨日报。

        窗口取 14 小时，覆盖"昨晚到今天早晨"的未读邮件，
        日报本身是只读的，不会触发任何邮箱动作。
        """
        plan, messages = self.plan_cleanup(hours=14, limit=200, custom_dimensions=custom_dimensions)
        return daily_brief(plan, messages)

    # ------------------------------------------------------------------ #
    # SOP 4：订阅费扫描
    # ------------------------------------------------------------------ #
    def scan_subscriptions(self) -> tuple[str, str]:
        """从最近 6 个月邮件里挖出订阅费。"""
        # 多个查询覆盖不同邮件提供方的描述词，互相补充
        queries = [
            '"subscription confirmation"',
            '"payment receipt"',
            '"renewal notice"',
            '"monthly charge"',
            '"your plan"',
            '"billing statement"',
        ]
        seen: dict[str, dict[str, object]] = {}
        for query in queries:
            for message in self.connector.search(query, months=6):
                row = self._extract_subscription(message)
                # 同一服务保留 last_charge_date 更新的一条，避免重复计数
                current = seen.get(str(row["service_name"]).lower())
                if current is None or row["last_charge_date"] > current["last_charge_date"]:
                    seen[str(row["service_name"]).lower()] = row
        rows = sorted(seen.values(), key=lambda row: str(row["service_name"]))
        # 写入历史用于将来"和上月对比"功能
        self.memory.subscription_history.append({"scanned_at": datetime.now().isoformat(), "items": rows})
        return subscriptions_markdown(rows), subscriptions_csv(rows)

    # ------------------------------------------------------------------ #
    # 辅助方法
    # ------------------------------------------------------------------ #
    def _category_for(self, plan: AgentPlan, email_id: str) -> Category | None:
        """在计划里反查某封邮件被归到哪一类。"""
        for judgment in plan.judgments:
            if judgment.email_id == email_id:
                return judgment.category
        return None

    def _extract_subscription(self, message: EmailMessage) -> dict[str, object]:
        """从一封订阅 / 账单邮件里抠出结构化字段。

        当前是粗暴的正则版本，足够 demo 用，生产版应改用 LLM 抽取。
        """
        text = f"{message.subject} {message.snippet} {message.body}"
        amount_match = re.search(r"\$([0-9]+(?:\.[0-9]{2})?)", text)
        amount = float(amount_match.group(1)) if amount_match else 0.0
        # 用域名前缀近似服务名，例如 billing@notion.so -> Notion
        service = message.sender.split("@")[-1].split(".")[0].title()
        plan_match = re.search(r"(Premium|Plus|Pro|Basic|Team|Enterprise)", text, re.IGNORECASE)
        return {
            "service_name": service,
            "monthly_amount": amount,
            "currency": "USD",
            "billing_cycle": "monthly",
            "last_charge_date": message.date.date().isoformat(),
            "plan_name": plan_match.group(1).title() if plan_match else "Unknown",
            "unsubscribe_link": "",
            "category": self._subscription_category(service),
        }

    def _subscription_category(self, service: str) -> str:
        """为服务名打一个粗略的分类标签，供报告分组使用。"""
        service_lower = service.lower()
        if service_lower in {"netflix", "spotify", "hulu"}:
            return "entertainment"
        if service_lower in {"notion", "github", "slack"}:
            return "productivity"
        return "other"


__all__ = ["MailTidyAgent"]

File: src/agent/legacy.ts
/**
 * 遗留的"流水线 SOP 编排"实现（Python 版 mailtidy/agent/legacy.py 的直译）。
 *
 * 这是迁移期的兼容层：让 demo 在 Phase 1 主循环（agent/loop.ts）落地之前
 * 还能跑。所有 SOP 都是简单的 fetch → classify → policy → execute → report
 * 流水线，没有 ReAct，没有主动调查，没有 checkpoint。
 *
 * Phase 1 完成后这个文件会被 `agent/loop.ts` 取代并删除。
 *
 * 四个 public 方法对应四条 SOP：
 *   - runCleanup       收件箱清理 + 报告
 *   - draftReplies     为 actionable 邮件批量草拟回复
 *   - dailyBriefing    早晨 2 分钟阅读的日报
 *   - scanSubscriptions 订阅费扫描 → Markdown + CSV
 */

import { ActionType, type AgentPlan, Category, type EmailMessage, type ExecutionResult, emptyExecutionResult } from "../data/models.js";
import { type AgentMemory, emptyMemory } from "../data/memory.js";
import {
  cleanupReport,
  dailyBrief,
  type SubscriptionRow,
  subscriptionsCsv,
  subscriptionsMarkdown,
} from "../data/reports.js";
import type { EmailConnector } from "../integrations/email/base.js";
import type { LLMClient } from "../llm/client.js";
import { DecisionPolicy } from "./policies.js";

export interface LegacyAgentOptions {
  connector: EmailConnector;
  llm: LLMClient;
  memory?: AgentMemory;
  policy?: DecisionPolicy;
}

export class LegacyMailTidyAgent {
  readonly memory: AgentMemory;
  private readonly connector: EmailConnector;
  private readonly llm: LLMClient;
  private readonly policy: DecisionPolicy;

  constructor(opts: LegacyAgentOptions) {
    this.connector = opts.connector;
    this.llm = opts.llm;
    this.memory = opts.memory ?? emptyMemory();
    this.policy = opts.policy ?? new DecisionPolicy();
  }

  async planCleanup(options: {
    hours?: number;
    limit?: number;
    customDimensions?: string[];
  } = {}): Promise<{ plan: AgentPlan; messages: EmailMessage[] }> {
    const messages = await this.connector.fetchRecent({
      hours: options.hours ?? 24,
      limit: options.limit ?? 200,
      unreadOnly: true,
    });
    const judgments = [];
    for (const message of messages) {
      const raw = await this.llm.classifyEmail(message, options.customDimensions);
      judgments.push(this.policy.applyMemory(raw, message.sender, this.memory));
    }
    return { plan: this.policy.buildPlan("inbox_cleanup", judgments), messages };
  }

  async executePlan(plan: AgentPlan, autoConfirm = false): Promise<ExecutionResult> {
    const result = emptyExecutionResult(plan.judgments.length);
    for (const action of plan.actions) {
      if (action.requiresConfirmation && !autoConfirm) {
        result.skippedConfirmation += action.emailIds.length;
        continue;
      }
      switch (action.action) {
        case ActionType.ARCHIVE:
          await this.connector.archive(action.emailIds);
          result.archived += action.emailIds.length;
          break;
        case ActionType.LABEL:
          if (action.label) {
            await this.connector.label(action.emailIds, action.label);
            result.labeled += action.emailIds.length;
          }
          break;
        case ActionType.STAR:
          await this.connector.star(action.emailIds);
          result.starred += action.emailIds.length;
          break;
        case ActionType.MARK_READ:
          await this.connector.markRead(action.emailIds);
          result.markedRead += action.emailIds.length;
          break;
        default:
          break;
      }
    }
    return result;
  }

  async runCleanup(options: {
    hours?: number;
    limit?: number;
    customDimensions?: string[];
    autoConfirm?: boolean;
  } = {}): Promise<string> {
    const { plan, messages } = await this.planCleanup(options);
    const result = await this.executePlan(plan, options.autoConfirm);
    const newsletters = messages.filter((m) => this.categoryFor(plan, m.id) === Category.NEWSLETTER);
    const summary = await this.llm.summarizeNewsletters(newsletters);
    return cleanupReport(plan, result, messages, summary);
  }

  async draftReplies(emailIds?: string[]): Promise<ExecutionResult> {
    const { plan, messages } = await this.planCleanup({ hours: 24, limit: 200 });
    const actionable = messages.filter(
      (m) =>
        this.categoryFor(plan, m.id) === Category.ACTIONABLE &&
        (emailIds === undefined || emailIds.includes(m.id)),
    );
    const result = emptyExecutionResult(actionable.length);
    for (const message of actionable) {
      const draft = await this.llm.draftReply(message, this.memory.styleProfile);
      await this.connector.saveDraft(message.id, draft);
      result.draftsCreated += 1;
    }
    return result;
  }

  async dailyBriefing(customDimensions?: string[]): Promise<string> {
    const opts: { hours: number; limit: number; customDimensions?: string[] } = {
      hours: 14,
      limit: 200,
    };
    if (customDimensions !== undefined) opts.customDimensions = customDimensions;
    const { plan, messages } = await this.planCleanup(opts);
    return dailyBrief(plan, messages);
  }

  async scanSubscriptions(): Promise<{ markdown: string; csv: string }> {
    const queries = [
      '"subscription confirmation"',
      '"payment receipt"',
      '"renewal notice"',
      '"monthly charge"',
      '"your plan"',
      '"billing statement"',
    ];
    const seen = new Map<string, SubscriptionRow>();
    for (const query of queries) {
      for (const message of await this.connector.search(query, 6)) {
        const row = this.extractSubscription(message);
        const key = row.serviceName.toLowerCase();
        const current = seen.get(key);
        if (!current || row.lastChargeDate > current.lastChargeDate) {
          seen.set(key, row);
        }
      }
    }
    const rows = [...seen.values()].sort((a, b) => a.serviceName.localeCompare(b.serviceName));
    this.memory.subscriptionHistory.push({
      scannedAt: new Date().toISOString(),
      items: rows as unknown as Record<string, unknown>[],
    });
    return { markdown: subscriptionsMarkdown(rows), csv: subscriptionsCsv(rows) };
  }

  private categoryFor(plan: AgentPlan, emailId: string): Category | undefined {
    return plan.judgments.find((j) => j.emailId === emailId)?.category;
  }

  private extractSubscription(message: EmailMessage): SubscriptionRow {
    const text = `${message.subject} ${message.snippet} ${message.body ?? ""}`;
    const amountMatch = text.match(/\$([0-9]+(?:\.[0-9]{2})?)/);
    const amount = amountMatch?.[1] ? parseFloat(amountMatch[1]) : 0;
    const domain = message.sender.split("@")[1] ?? "";
    const service = (domain.split(".")[0] ?? "Unknown").replace(/^./, (c) => c.toUpperCase());
    const planMatch = text.match(/Premium|Plus|Pro|Basic|Team|Enterprise/i);
    return {
      serviceName: service,
      monthlyAmount: amount,
      currency: "USD",
      billingCycle: "monthly",
      lastChargeDate: message.date.slice(0, 10),
      planName: planMatch ? planMatch[0].replace(/^./, (c) => c.toUpperCase()) : "Unknown",
      category: this.subscriptionCategory(service),
    };
  }

  private subscriptionCategory(service: string): string {
    const s = service.toLowerCase();
    if (["netflix", "spotify", "hulu"].includes(s)) return "entertainment";
    if (["notion", "github", "slack"].includes(s)) return "productivity";
    return "other";
  }
}
